### PR TITLE
Implementing Script of the Ammo Vendors Extension

### PR DESCRIPTION
**Pull request by JohnnyPlayy**
* **Addressed Issue(s)**: None

* **Server Mode**:  Renewal

* **Description of Pull Request**: 

image

Thanks to @Lemongrass3110 for developing the BarterShop

TODOs:

- [ ] Add the correct NPC rotation

### CODE REVIEW

File: npc/re/scripts_athena.conf [hunk 1]
```diff
@@ -117,6 +117,7 @@ npc: npc/re/merchants/enchan_mal.txt
 npc: npc/re/merchants/enchan_mora.txt
 npc: npc/re/merchants/enchan_rockridge.txt
 npc: npc/re/merchants/enchan_verus.txt
+npc: npc/re/merchants/Extended_Ammunition.txt.txt
```
Comment (Everade) [inline npc/re/scripts_athena.conf:120]: npc: npc/re/merchants/Extended_Ammunition.txt

**Review by Lemongrass3110 — CHANGES_REQUESTED** [review]

**Comment by Lemongrass3110** [comment]
Reported by CI:
```
[Status]: Loading 'npc/barters.yml'...
Error: : Failed to read BARTER_DB database file from 'npc/re/merchants/barters.yml'.
[Status]: Loading 'npc/re/merchants/barters.yml'...
Error: : illegal map value (Line 147: Column 18)
[Status]: Loading 'npc/custom/barters.yml'...
```

File: npc/re/merchants/barters.yml [hunk 2]
```diff
@@ -44,3 +44,350 @@
 Header:
   Type: BARTER_DB
   Version: 1
+Body:
```
Comment (Atemo) [inline npc/re/merchants/barters.yml:47]: How about 
```
Footer:
  Imports:
  - Path: npc/re/merchants/barters/Extended_Ammunition.yml
```
to organize the shops?

File: npc/re/merchants/shops.txt [hunk 3]
```diff
@@ -51,8 +51,8 @@ dewata,158,182,6	shop	Armor Dealer	536,2211:-1,2401:-1,2403:-1,2501:-1,2503:-1,2
 //=======================================================
 // Einbroch
 //=======================================================
-einbroch,122,250,4	shop	Black Marketeer#ein	49,2139:-1,2800:-1,2801:-1,2802:-1,2803:-1,2804:-1,2805:-1,2806:-1,2807:-1,18000:-1,18001:-1,18002:-1,18003:-1,18004:-1,12392:-1,12393:-1,12394:-1,6145:-1,6146:-1,6147:-1,6186:-1,2808:-1
-que_ng,180,79,3	shop	Johnny Waiker	900,13200:-1,13221:-1,13222:-1
+//einbroch,122,250,4	shop	Black Marketeer#ein	49,2139:-1,2800:-1,2801:-1,2802:-1,2803:-1,2804:-1,2805:-1,2806:-1,2807:-1,18000:-1,18001:-1,18002:-1,18003:-1,18004:-1,12392:-1,12393:-1,12394:-1,6145:-1,6146:-1,6147:-1,6186:-1,2808:-1
```
[the remark this answers is not in the document]
Comment (JohnnyPlayy) [reply npc/re/merchants/shops.txt:54]: I thought it was a really good idea! But do you want me to add a feature that toggles the old ammo NPCs on and off?

[the remark this answers is not in the document]
Comment (JohnnyPlayy) [reply npc/re/merchants/shops.txt:54]: Like this example?

//ON: enable ammo extensions and disable old ammo dealers
//OFF: disables the minis extension and enables the old ammo traders

feature.extension_ammunition: on

[the remark this answers is not in the document]
Comment (Lemongrass3110) [reply npc/re/merchants/shops.txt:54]: I think we need to add general "feature.barter" and "feature.barter_ex".
Just leave the "if" for now and add it without if. :-)

**Comment by RagnaWay** [comment]
Wasn't the marketshop items supposed to be unlimited?

**Comment by Lemongrass3110** [comment]
> Wasn't the marketshop items supposed to be unlimited?

Not supported by rAthena (yet). Not his fault.

File: npc/re/merchants/barters/Extended_Ammunition.yml [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+            Item: Special_Alloy_Trap
+            Amount: 500
+
+###########################################################################
+##= Ammunition Trader Shop
+###########################################################################
+  - Name: Box_Exchange_Gunslinger
+    Items:
+      - Index: 0
+        Item: Bullet_Case
```
Comment (RagnaWay) [inline npc/re/merchants/barters/Extended_Ammunition.yml:182]: I don't think I can go wrong, but delete this space after **Bullet_Case**

**Comment by RagnaWay** [comment]
```[Error]: npc_market_checkall_sub: NPC 'Ammunition_Gunslinger' is not proper for market, removing...
[Error]: npc_market_checkall_sub: NPC 'Ammunition_Mechanic' is not proper for market, removing...
[Error]: npc_market_checkall_sub: NPC 'Weapons_Mechanic' is not proper for market, removing...
[Error]: npc_market_checkall_sub: NPC 'Weapons_Ninja' is not proper for market, removing...
[Error]: npc_market_checkall_sub: NPC 'Weapons_Gunslinger' is not proper for market, removing...
[Error]: npc_market_checkall_sub: NPC 'Talisman_Merchant' is not proper for market, removing...
[Error]: npc_market_checkall_sub: NPC 'Arrow_Archer' is not proper for market, removing...

**Comment by JohnnyPlayy** [comment]
> ```
> [Error]: npc_market_checkall_sub: NPC 'Ammunition_Mechanic' is not proper for market, removing...
> [Error]: npc_market_checkall_sub: NPC 'Weapons_Mechanic' is not proper for market, removing...
> [Error]: npc_market_checkall_sub: NPC 'Weapons_Ninja' is not proper for market, removing...
> [Error]: npc_market_checkall_sub: NPC 'Weapons_Gunslinger' is not proper for market, removing...
> [Error]: npc_market_checkall_sub: NPC 'Talisman_Merchant' is not proper for market, removing...
> [Error]: npc_market_checkall_sub: NPC 'Arrow_Archer' is not proper for market, removing...
> ```

How I changed the marketshop to shop! Turns out the rathena recognizes this.

**Comment by RagnaWay** [comment]
> > ```
> > [Error]: npc_market_checkall_sub: NPC 'Ammunition_Mechanic' is not proper for market, removing...
> > [Error]: npc_market_checkall_sub: NPC 'Weapons_Mechanic' is not proper for market, removing...
> > [Error]: npc_market_checkall_sub: NPC 'Weapons_Ninja' is not proper for market, removing...
> > [Error]: npc_market_checkall_sub: NPC 'Weapons_Gunslinger' is not proper for market, removing...
> > [Error]: npc_market_checkall_sub: NPC 'Talisman_Merchant' is not proper for market, removing...
> > [Error]: npc_market_checkall_sub: NPC 'Arrow_Archer' is not proper for market, removing...
> > ```
> 
> How I changed the marketshop to shop! Turns out the rathena recognizes this.

after changing to 'shop', you added ``, 1;`` I think it doesn't work for that type.
```
callshop "Arrow_Archer", 1;

File: npc/re/merchants/barters.yml [hunk 5]
```diff
@@ -44,3 +44,7 @@
 Header:
   Type: BARTER_DB
   Version: 1
+
+Footer:
+  Imports:
+  - Path: npc/re/merchants/barters/Extended_Ammunition.yml
```
Comment (Atemo) [inline npc/re/merchants/barters.yml:50]: Add a newline at the end of the file
```suggestion
  - Path: npc/re/merchants/barters/Extended_Ammunition.yml

```

File: npc/re/merchants/Extended_Ammunition.txt [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+OnInit:
+    setunitdata getnpcid(0),UNPC_GROUP_ID,12;
+	end;
+}
+
+//= Basic Shops
+//============================================================
+
+//= Archer
+-	shop	Arrow_Archer	HIDDEN_NPC,1750:-1:1,1770:-1:1,1751:-1:1,1752:-1:1,1754:-1:1,1755:-1:1,1756:-1:1,1757:-1:1,1762:-1:1,1065:-1:1,7940:-1:1
```
Comment (Atemo) [inline npc/re/merchants/Extended_Ammunition.txt:196]: The definition is wrong and must be updated for the normal shop
normal: `<itemid>:<price>`
market: `<itemid>:<price>:<quantity>`

Btw what is the difference between an unlimited marketshop and a normal shop?

Comment (JohnnyPlayy) [reply npc/re/merchants/Extended_Ammunition.txt:196]: I forgot about it! Yesterday I posted this at bedtime. Very tired. It turned out that I wrote without checking.

File: npc/re/merchants/Extended_Ammunition.txt [hunk 7]
```diff
@@ -8,6 +8,52 @@
 // Texts from this script were translated according to KRO
 //============================================================
 
+//= NPCs that become useless when activating the batashop!
```
Comment (RagnaWay) [inline npc/re/merchants/Extended_Ammunition.txt:11]: batashop = Barter Shop

**Pull request closed without merging**